### PR TITLE
Side dish recommendation + Performance improvement

### PR DESCRIPTION
**Pull request by erikzimmermann**

### CODE REVIEW

**Comment by erikzimmermann** [comment]
Closes #191

File: backend/mensa_recommend/source/recommendation/recommender.py [hunk 1]
```diff
@@ -448,26 +461,41 @@ def __compute_mensa_distance_scores(self) -> Dict[date, Dict[int, float]]:
...
+            plans = DishPlan.objects.filter(date=day, dish__main=True) \
+                .prefetch_related("dish", "mensa", "dish__categories",
+                                  "dish__allergies")
+
+            for plan in plans:
+                did = plan.dish_id
+                score = self.__dist_to_mensa(day, plan.mensa_id)
+
+                current = dish_to_plan.get(did)
+                if current is None or score > current[1]:
```
Comment (martenjostmann) [inline backend/mensa_recommend/source/recommendation/recommender.py:493]: Shouldn't it be score < current[1] ?

File: backend/mensa_recommend/source/recommendation/recommender.py [hunk 2]
```diff
@@ -448,26 +461,41 @@ def __compute_mensa_distance_scores(self) -> Dict[date, Dict[int, float]]:
...
             for key in distances.keys():
                 score = 1 - min(distances[key] / self._flexibility, 1)
                 distances[key] = score
 
             week_distances[day] = distances
         return week_distances
 
-    def __load_dish_plan(self) -> Dict[date, List[DishPlan]]:
+    def __load_dish_plan(self) -> Dict[date, Dict[int, DishPlan]]:
         """Load the dish plan from the database.
```
Comment (martenjostmann) [inline backend/mensa_recommend/source/recommendation/recommender.py:473]: Some more comments would be nice. When I understand the function right it is not just loading dishes but also filtering same dishes with higher distance to the mensa

File: backend/mensa_recommend/source/processing/conditions.py [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+
+class TimeCondition(Condition):
+    def __init__(self, lifetime: timedelta):
+        self._lifetime = lifetime.total_seconds()
+        self._timestamp: Optional[datetime] = None
+
+    def holds(self) -> bool:
+        now = datetime.now()
+        result = self._timestamp is not None and (now - self._timestamp) \
+            .seconds < self._lifetime
```
Comment (martenjostmann) [inline backend/mensa_recommend/source/processing/conditions.py:30]:
```suggestion
        result = self._timestamp is not None and 
        (now - self._timestamp).seconds < self._lifetime
```

File: backend/mensa_recommend/source/recommendation/recommender.py [hunk 4]
```diff
@@ -86,6 +52,66 @@ def multiply(f: float, factors: List[Optional[float]],
...
+            self._cache[broad_zip] = weather_data
+
+        data: Optional[Tuple[TimeCondition, float]] = weather_data.get(t)
+        if data is None:
+            weather_data[t] = TimeCondition(self._lifetime), \
+                weather.get_score(t, mensa.lat, mensa.lon)
+        elif not data[0].holds():
+            weather_data[t] = data[0], weather.get_score(t, mensa.lat,
+                                                         mensa.lon)
+        return self._cache[broad_zip][t][1]
```
Comment (martenjostmann) [inline backend/mensa_recommend/source/recommendation/recommender.py:112]: More comments would be nice for better readability

File: backend/mensa_recommend/source/recommendation/recommender.py [hunk 5]
```diff
@@ -95,9 +121,8 @@ class DishRecommender:
         Disclaimer: This should not be used in production but rather for a
         first demo and a most viable product.
```
Comment (martenjostmann) [inline backend/mensa_recommend/source/recommendation/recommender.py:122]: I think we can remove this now.

File: backend/mensa_recommend/source/recommendation/recommender.py [hunk 6]
```diff
@@ -95,9 +121,8 @@ class DishRecommender:
...
     """
-    # This version of a sentence transformer was the smallest I could find.
-    # Size: ~100MB
-    sentence_transformer = SentenceTransformer('all-MiniLM-L6-v2')
+    _dish_encoder = DishEncoderCache()
+    _weather = WeatherCache()
```
Comment (martenjostmann) [inline backend/mensa_recommend/source/recommendation/recommender.py:125]: Maybe a comment describing where these static variables are for

File: backend/requirements.txt [hunk 7]
```diff
@@ -1,15 +1,16 @@
...
+redis==4.3.4
+tqdm==4.64.1
+django-cors-headers==3.13.0
+djangorestframework==3.14.0
+djangorestframework-simplejwt==5.2.2
+Google-Images-Search==1.4.6
+pycryptodome==3.14.1
+numpy==1.23.5
+sentence_transformers==2.2.2
+joblib==1.2.0
```
Comment (martenjostmann) [inline backend/requirements.txt:16]: is joblib used right now?

**Review by martenjostmann — CHANGES_REQUESTED** [review]

File: backend/mensa_recommend/source/recommendation/recommender.py (hunk 1 above)
Comment (erikzimmermann) [reply backend/mensa_recommend/source/recommendation/recommender.py:493]: I had it before and the score is between 0..1. The higher the better.

**Review by martenjostmann — APPROVED** [review]

**Pull request merged**